src/test/mir-opt/const_prop/usize_literal_propagation/rustc.main.ConstProp.diff

```diff
@@ -0,0 +1,58 @@
+- // MIR for `main` before ConstProp
++ // MIR for `main` after ConstProp
+  
+  fn main() -> () {
+      let mut _0: ();                      // return place in scope 0 at $DIR/usize_literal_propagation.rs:2:11: 2:11
+      let _2: &i32;                        // in scope 0 at $DIR/usize_literal_propagation.rs:5:9: 5:46
+      let mut _3: usize;                   // in scope 0 at $DIR/usize_literal_propagation.rs:5:44: 5:45
+      scope 1 {
+          let _1: usize;                   // in scope 1 at $DIR/usize_literal_propagation.rs:4:13: 4:14
+          scope 2 {
+              debug x => _1;               // in scope 2 at $DIR/usize_literal_propagation.rs:4:13: 4:14
+          }
+      }
+  
+      bb0: {
+          StorageLive(_1);                 // bb0[0]: scope 1 at $DIR/usize_literal_propagation.rs:4:13: 4:14
+          _1 = const 0usize;               // bb0[1]: scope 1 at $DIR/usize_literal_propagation.rs:4:17: 4:18
+                                           // ty::Const
+                                           // + ty: usize
+                                           // + val: Value(Scalar(0x0000000000000000))
+                                           // mir::Constant
+                                           // + span: $DIR/usize_literal_propagation.rs:4:17: 4:18
+                                           // + literal: Const { ty: usize, val: Value(Scalar(0x0000000000000000)) }
+          StorageLive(_2);                 // bb0[2]: scope 2 at $DIR/usize_literal_propagation.rs:5:9: 5:46
+          StorageLive(_3);                 // bb0[3]: scope 2 at $DIR/usize_literal_propagation.rs:5:44: 5:45
+-         _3 = _1;                         // bb0[4]: scope 2 at $DIR/usize_literal_propagation.rs:5:44: 5:45
++         _3 = const 0usize;               // bb0[4]: scope 2 at $DIR/usize_literal_propagation.rs:5:44: 5:45
++                                          // ty::Const
++                                          // + ty: usize
++                                          // + val: Value(Scalar(0x0000000000000000))
++                                          // mir::Constant
++                                          // + span: $DIR/usize_literal_propagation.rs:5:44: 5:45
++                                          // + literal: Const { ty: usize, val: Value(Scalar(0x0000000000000000)) }
+          _2 = const std::intrinsics::transmute::<usize, &i32>(move _3) -> bb1; // bb0[5]: scope 2 at $DIR/usize_literal_propagation.rs:5:9: 5:46
```

Comment:
I think we should be seeing this right?

```suggestion
          _2 = const std::intrinsics::transmute::<usize, &i32>(const 0usize) -> bb1; // bb0[5]: scope 2 at $DIR/usize_literal_propagation.rs:5:9: 5:46
```

Reply:
I think this FIXME needs to be resolved for that to happen

https://github.com/rust-lang/rust/blob/8001b96f48d2c9cd646acd2d16f4e06078f258e1/src/librustc_mir/transform/const_prop.rs#L968-L969

Reply:
Ohhh! That's a really good idea :D Imma try it!

Reply:
So, if I got this right... to do this, I need to:

(Note: excessive `continue`s are of course not necessary, but I added them for clarity)

1) For each argument `arg` in `TerminatorKind::Call { args, .. }`, I do:
    1) Check if `arg` should be const-propagated.
    2) If **no**, `continue`
    3) Check if `eval_operand` gives `Some(value)`.
    4) If **no**, `continue`
    5) Do `*arg = value`
2) Done?

I think that's it. Right?
If I'm understanding it correctly, in the const-prop step we visit each `Operand`, and check if its value could've been propagated from somewhere above it in the MIR. 

If that is the case, then what we're trying to do for this particular FIXME is that we're trying to figure out, while visiting a function, if its arguments' values could've been const-propagated from above. Am I right?

@wesleywiser

Reply:
That sounds correct to me!

Reply:
I think I've implemented this now! It further optimized (or so it seemed to me) another `mir-opt` test. It also gives the output we're looking for for this test now!

However... it seems that now it doesn't pass a good chunk of `compiletest`'s tests. We'll have to see why, I hope I didn't break something but it's very likely that I did. I left that work for tomorrow :3